File: go/triangle/triangle.go
package triangle

import (
	"math"
	"sort"
)

// KindFromSides determines if a triangle is equilateral, isosceles, or
// scalene.
func KindFromSides(a, b, c float64) Kind {
	if math.IsNaN(a) || math.IsNaN(b) || math.IsNaN(c) {
		return NaT
	}

	sides := []float64{a, b, c}
	sort.Sort(sort.Float64Slice(sides))
	if sides[0] <= 0 || sides[0]+sides[1] <= sides[2] {
		return NaT
	}

	if a == b && b == c {
		return Equ
	} else if a == b || b == c || a == c {
		return Iso
	} else {
		return Sca
	}
}

// Kind represents a type of triangle.
type Kind int8

// Constants representing the the types of triangle.
const (
	Equ Kind = 0
	Iso Kind = 1
	Sca Kind = 2
	NaT Kind = 3
)

File: ruby/triangle/triangle.rb
TriangleError = Class.new(StandardError)

class Triangle
  attr_reader :sides

  def initialize(*sides)
    @sides = sides.sort
  end

  def kind
    a, b, c = sides

    raise TriangleError if a <= 0 || a + b <= c

    if sides.uniq.size == 1
      :equilateral
    elsif sides.uniq.size == 2
      :isosceles
    else
      :scalene
    end
  end
end
